### ISSUE TITLE
bash autocompletion for nix shell (nix command)

### ISSUE THREAD

**Issue by roberth** [NixOS/nix#4453]
**Is your feature request related to a problem? Please describe.**

It's frustrating when autocompletion doesn't work in a shell.

**Describe the solution you'd like**

Simple option: add entries to `XDG_DATA_DIRS`.

Complicated option: make `nix shell` use stdenv, which can take care of this logic, at the cost of a Nixpkgs dependency.

**Describe alternatives you've considered**

Write GUIs for everything that shouldn't have a GUI.

**Additional context**

Completions are available for derivation-based shells since https://github.com/NixOS/nixpkgs/pull/103501

`nix-shell` also needs adjustment, see #4452

**Comment by roberth** [comment]
Another way to make this work is by using `passthru` to inform how a package should added to the environment variables. See also [`env` noun suggestion](https://github.com/NixOS/nix/issues/4715#issuecomment-875729369)

This is similar to what the `stdenv` machinery does in a package build, but simplified and specifically for use _outside_ a package build.
Perhaps this should be incorporated into the "app" concept or vice versa.

**Comment by roberth** [comment]
Basically what I'm suggesting is to treat "environment variables" as a serious deployment target, where a package can tell Nix exactly what to do, or rely on the default (`PATH` + possibly `XDG_DATA_DIRS`)

Something like

```nix
passthru.useViaEnvironment = {
  PATH = "${self.outPath}/bin:${commonlyUsedRuntimeDep}/bin";
  XDG_DATA_DIRS = "${self.outPath}/share/foo";
  shellHook = "echo To initialize the foo package config, run foo init";
}
```

just to illustrate the concept.

I'm assuming https://github.com/NixOS/nixpkgs/pull/119942 for simplicity on the Nix side of this interface. The self-referential nature of this feature can be worked around with placeholders, but I would prefer to exploit the existing language instead.

How the exactly the environment variables should be merged isn't obvious. Nix could hardcode some of them. Nixpkgs doesn't know how to do it yet either https://github.com/NixOS/nixpkgs/issues/122244.

**Comment by SuperSandro2000** [comment]
Is this similar to https://github.com/nix-community/lorri/pull/61 ?

**Comment by roberth** [comment]
@SuperSandro2000 yes, it's all necessary to make completions work beyond simple nix-shell + stdenv (https://github.com/NixOS/nixpkgs/pull/103501)

**Comment by roberth** [comment]
`nix shell` (or whatever it will be called #4715) could solve this by depending on https://github.com/NixOS/nixpkgs/pull/132617, allowing a small amount of behavior to be added, without burdening Nix itself with it.
